### PR TITLE
feat(iota-core): add deny rule proposal persistence, stake aggregation, and active-set cache (#12151)

### LINKED ISSUE #12151: Add deny rule proposal persistence, stake aggregation, and active-set cache
Part of #10749. Adds the data layer for deny governance: per-authority proposal storage, stake-weighted aggregation into the active rule set, and a crash-safe in-memory cache. No consensus wiring yet — unit-tested in isolation (mirrors the `OverloadNotificationV1` pattern).

### PR DESCRIPTION
**Pull request by filipdulic**
# Description of change

Adds the data layer for deny governance (#10749), mirroring the `OverloadNotificationV1` pattern: per-authority proposal storage, stake-weighted aggregation into the active rule set, and a crash-safe in-memory cache. No consensus wiring yet — nothing records proposals in production (the recording entry point is `#[cfg(test)]` until the consensus handler lands), so the change is behavior-neutral and unit-tested in isolation.

- **Persistence:** epoch-scoped `deny_rule_proposals: DBMap<AuthorityName, DenyRuleProposal>`; proposals recorded into `ConsensusCommitOutput` and flushed atomically with `last_consensus_stats`; quarantine cache seeded at construction, advanced in lockstep at flush, overlaid with queued commits by `current_deny_rule_proposals()`.
- **Generation dedup:** a proposal supersedes only with a strictly newer generation, both across commits (`should_record_deny_rule_proposal`, backed by a no-clone `deny_rule_proposal_generation` lookup) and within one commit (`ConsensusCommitOutput::record_deny_rule_proposal`).
- **Aggregation:** `compute_active_deny_rules(proposals, committee)` — pure function; deny-list entries activate at f+1 (`validity_threshold`), kill switches at 2f+1 (`quorum_threshold`); non-members weigh 0. Each switch is tallied through a named field accessor (no positional coupling).
- **Active set:** `active_deny_rules: ArcSwap<DenyRuleSet>` + `get_active_deny_rules()`; seeded in `new()` by recomputing from the persisted table, so a mid-epoch restart resumes with the rules from all flushed commits (unflushed commits are re-recorded by consensus replay). `update_active_deny_rules()` reports whether the set changed — the seam a future on-chain materialization attaches to.
- `EpochStartConfiguration` (+V1/V2) gains `Clone`, enabling same-epoch store reconstruction in tests.

## Links to any relevant issues

Fixes #12151.

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration tests)
- [x] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by bingyanglin — APPROVED** [review]

**Review by roman1e2f5p8s — APPROVED** [review]

File: crates/iota-core/src/authority/authority_per_epoch_store.rs [hunk 1]
```diff
@@ -2943,6 +2970,113 @@ impl AuthorityPerEpochStore {
...
+        info!(?rules, "active deny rules changed");
+        self.active_deny_rules.store(Arc::new(rules));
+        Ok(true)
+    }
+
+    /// Computes the stake-weighted aggregate of the given deny rule proposals:
+    /// a deny list entry is active with f+1 supporting stake, a kill switch
+    /// with 2f+1. Pure function of the proposals and committee, so every
+    /// validator derives the same set from the same consensus state.
+    pub(crate) fn compute_active_deny_rules(
```
Comment (piotrm50) [inline crates/iota-core/src/authority/authority_per_epoch_store.rs]: This function is key to the logic but it's only covered by a single test with out hitting edge cases for stake thresholds. Also, not all types of rules are covered by the test.

File: crates/iota-core/src/authority/authority_per_epoch_store.rs [hunk 2]
```diff
@@ -2943,6 +2970,113 @@ impl AuthorityPerEpochStore {
...
+
+    /// Whether `proposal` is newer than the recorded proposal (if any) from
+    /// the same authority and should therefore be recorded.
+    pub fn should_record_deny_rule_proposal(&self, proposal: &DenyRuleProposal) -> bool {
+        let stale = self
+            .consensus_quarantine
+            .read()
+            .deny_rule_proposal_generation(&proposal.authority)
+            .is_some_and(|generation| generation >= proposal.generation);
+        if stale {
```
Comment (piotrm50) [inline crates/iota-core/src/authority/authority_per_epoch_store.rs]: does this log belong here or in the caller?

File: crates/iota-core/src/authority/authority_per_epoch_store.rs [hunk 3]
```diff
@@ -698,6 +699,11 @@ pub struct AuthorityPerEpochStore {
     /// `Rejected` response instead of waiting for the gRPC deadline.
     dropped_tx_status_cache: super::dropped_tx_status_cache::DroppedTxStatusCache,
 
+    /// The active deny rule set: the stake-weighted aggregate of the recorded
+    /// deny rule proposals. Recomputed via `update_active_deny_rules` and
+    /// seeded from the `deny_rule_proposals` table on construction.
+    active_deny_rules: ArcSwap<DenyRuleSet>,
```
Comment (piotrm50) [inline crates/iota-core/src/authority/authority_per_epoch_store.rs]: shoul we call it transaction_deny_rules to contrast and avoid confusion with coin_deny_rules?

File: crates/iota-core/src/authority/epoch_start_configuration.rs [hunk 4]
```diff
@@ -98,7 +98,7 @@ impl fmt::Display for EpochFlag {
 }
 
 /// Parameters of the epoch fixed at epoch start.
-#[derive(Serialize, Deserialize, Debug, Eq, PartialEq)]
+#[derive(Clone, Serialize, Deserialize, Debug, Eq, PartialEq)]
```
Comment (piotrm50) [inline crates/iota-core/src/authority/epoch_start_configuration.rs:101]: Why do we need clone here?

**Review by piotrm50 — APPROVED** [review]
left some nit comments, approve overall

File: crates/iota-core/src/authority/epoch_start_configuration.rs (hunk 4 above)
Comment (filipdulic) [reply crates/iota-core/src/authority/epoch_start_configuration.rs:101]: clone is needed by the new persistence test, which reopens the epoch store to check that recorded proposals survive a restart: `AuthorityPerEpochStore::new` takes the config by value, while the original store holds it in an `Arc`. The enum is plain data, so the derive is cheap and safe.

File: crates/iota-core/src/authority/authority_per_epoch_store.rs (hunk 1 above)
Comment (filipdulic) [reply crates/iota-core/src/authority/authority_per_epoch_store.rs]: Extended the test to cover all three deny lists and all six switches at 3/2/1/0 supporting members plus a non-member. Note the lists share one generic activated() helper and the switches share switch_active, so code-path coverage was already complete — this makes field coverage explicit.

File: crates/iota-core/src/authority/authority_per_epoch_store.rs (hunk 2 above)
Comment (filipdulic) [reply crates/iota-core/src/authority/authority_per_epoch_store.rs]: moved to an else branch at the call site.

File: crates/iota-core/src/authority/authority_per_epoch_store.rs (hunk 3 above)
Comment (filipdulic) [reply crates/iota-core/src/authority/authority_per_epoch_store.rs]: Renamed across the stack: active_transaction_deny_rules + matching accessors.
